OrbitVulkanLayer/SubmissionTracker.h

```diff
@@ -0,0 +1,816 @@
+                                       .color = marker.color.value(),
+                                       .begin_info = submitted_marker,
+                                       .depth = markers.marker_stack.size(),
+                                       .depth_exceeds_maximum = marker.cut_off};
+              markers.marker_stack.push(std::move(marker_state));
+              break;
+            }
+
+            case MarkerType::kDebugMarkerEnd: {
+              MarkerState marker_state = markers.marker_stack.top();
+              markers.marker_stack.pop();
+
+              // If there is a begin marker slot from a previous submission, this is our chance to
+              // reset the slot.
+              if (marker_state.begin_info.has_value() && !queue_submission_optional.has_value()) {
+                marker_slots_to_reset.push_back(marker_state.begin_info.value().slot_index);
+              }
+
+              // If the begin marker was cut off, but the end marker was not (as it is in a
+              // different submission), we reset the slot
+              if (marker_state.depth_exceeds_maximum && marker.slot_index.has_value()) {
+                marker_slots_to_reset.push_back(marker.slot_index.value());
+              }
+
+              if (queue_submission_optional.has_value() && marker.slot_index.has_value() &&
+                  !marker_state.depth_exceeds_maximum) {
+                CHECK(submitted_marker.has_value());
+                queue_submission_optional->completed_markers.emplace_back(
+                    SubmittedMarkerSlice{.begin_info = marker_state.begin_info,
+                                         .end_info = submitted_marker.value(),
+                                         .depth = marker_state.depth,
+                                         .color = marker_state.color,
+                                         .label_name = std::move(marker_state.label_name)});
+              }
+
+              break;
+            }
+          }
+        }
+        command_buffer_to_state_.erase(command_buffer);
+      }
+    }
+
+    if (!marker_slots_to_reset.empty()) {
+      timer_query_pool_->ResetQuerySlots(device, marker_slots_to_reset);
+    }
+
+    if (!queue_submission_optional.has_value()) {
+      return;
+    }
+
+    if (!queue_to_submissions_.contains(queue)) {
+      queue_to_submissions_[queue] = {};
+    }
+    CHECK(queue_to_submissions_.contains(queue));
+    queue_to_submissions_.at(queue).emplace_back(std::move(queue_submission_optional.value()));
+  }
+
+  // This method is responsible to retrieve all the timestamps for the "completed" submissions,
+  // for transforming the information of those submissions (in particlular about the command buffers
+  // and debug markers) into the `GpuQueueSubmission` proto, and for sending it to the
+  // `VulkanLayerProducer`. We consider a submission to be "completed" iff its last command buffer
```

Comment:
Nit: Since this is a definition of "completed", I would use "if" and not "iff" (the "other" direction is true trivially by definition). 

```suggestion
  // `VulkanLayerProducer`. We consider a submission to be "completed" if its last command buffer
```

Reply:
Agree.